File: packages/runtime-protocol/src/terminal-multiplex/snapshot-records.ts
export type TerminalMultiplexSnapshotReason =
  | 'initial'
  | 'manual'
  | 'recovery'
  | 'reveal'
  | 'resume'
  | 'pending-cap'
  | 'normal-buffer-resize'

const SNAPSHOT_REASON_BY_WIRE = [
  'initial',
  'manual',
  'recovery',
  'reveal',
  'resume',
  'pending-cap',
  'normal-buffer-resize'
] as const satisfies readonly TerminalMultiplexSnapshotReason[]

export type TerminalMultiplexSnapshotStartRecord = {
  snapshotId: number
  reason: TerminalMultiplexSnapshotReason
  source: 0 | 1
  activeBuffer: 0 | 1
  truncated: boolean
  byteBudget: boolean
  coldRestore: boolean
  cols: number
  rows: number
  retainedScrollbackRows: number
  coverageEndSeq: bigint
  pendingDeliveryStartSeq: bigint
  sectionBytes: readonly [number, number, number, number, number]
}

export type TerminalMultiplexSnapshotChunkRecord = {
  snapshotId: number
  section: 0 | 1 | 2 | 3 | 4
  sectionOffset: number
  data: Uint8Array<ArrayBufferLike>
}

export type TerminalMultiplexSnapshotEndRecord = {
  snapshotId: number
  status: 0 | 1 | 2 | 3
  coverageEndSeq: bigint
  assembledBytes: number
  crc32c: number
}

export const TERMINAL_MULTIPLEX_SNAPSHOT_RECORD_WIRE = {
  start: {
    bytes: 64,
    snapshotIdOffset: 0,
    reasonOffset: 4,
    reasonMax: 6,
    sourceOffset: 5,
    sourceMax: 1,
    activeBufferOffset: 6,
    activeBufferMax: 1,
    flagsOffset: 7,
    flagsMask: 7,
    truncatedFlag: 1,
    byteBudgetFlag: 2,
    coldRestoreFlag: 4,
    colsOffset: 8,
    colsMin: 1,
    colsMax: 1000,
    rowsOffset: 10,
    rowsMin: 1,
    rowsMax: 500,
    retainedScrollbackRowsOffset: 12,
    reserved64Offset: 16,
    coverageEndSeqOffset: 24,
    pendingDeliveryStartSeqOffset: 32,
    sectionBytesOffset: 40,
    sectionCount: 5,
    sectionStrideBytes: 4,
    reserved32Offset: 60
  },
  chunk: {
    headerBytes: 16,
    snapshotIdOffset: 0,
    sectionOffset: 4,
    sectionMax: 4,
    reserved8Offset: 5,
    reserved16Offset: 6,
    dataOffsetOffset: 8,
    dataBytesOffset: 12,
    maxDataBytes: 48 * 1024
  },
  end: {
    bytes: 24,
    snapshotIdOffset: 0,
    statusOffset: 4,
    statusMax: 3,
    reserved8Offset: 5,
    reserved16Offset: 6,
    coverageEndSeqOffset: 8,
    assembledBytesOffset: 16,
    crc32cOffset: 20
  }
} as const

export const TERMINAL_MULTIPLEX_SNAPSHOT_CHUNK_DATA_BYTES =
  TERMINAL_MULTIPLEX_SNAPSHOT_RECORD_WIRE.chunk.maxDataBytes

export function encodeTerminalMultiplexSnapshotStartRecord(
  record: TerminalMultiplexSnapshotStartRecord
): Uint8Array<ArrayBuffer> {
  const wire = TERMINAL_MULTIPLEX_SNAPSHOT_RECORD_WIRE.start
  const out = new Uint8Array(wire.bytes)
  const view = new DataView(out.buffer)
  const flags =
    (record.truncated ? wire.truncatedFlag : 0) |
    (record.byteBudget ? wire.byteBudgetFlag : 0) |
    (record.coldRestore ? wire.coldRestoreFlag : 0)
  view.setUint32(wire.snapshotIdOffset, record.snapshotId, true)
  view.setUint8(wire.reasonOffset, encodeTerminalMultiplexSnapshotReason(record.reason))
  view.setUint8(wire.sourceOffset, record.source)
  view.setUint8(wire.activeBufferOffset, record.activeBuffer)
  view.setUint8(wire.flagsOffset, flags)
  view.setUint16(wire.colsOffset, record.cols, true)
  view.setUint16(wire.rowsOffset, record.rows, true)
  view.setUint32(wire.retainedScrollbackRowsOffset, record.retainedScrollbackRows, true)
  view.setBigUint64(wire.reserved64Offset, 0n, true)
  view.setBigUint64(wire.coverageEndSeqOffset, record.coverageEndSeq, true)
  view.setBigUint64(wire.pendingDeliveryStartSeqOffset, record.pendingDeliveryStartSeq, true)
  record.sectionBytes.forEach((bytes, index) =>
    view.setUint32(wire.sectionBytesOffset + index * wire.sectionStrideBytes, bytes, true)
  )
  view.setUint32(wire.reserved32Offset, 0, true)
  return out
}

export function decodeTerminalMultiplexSnapshotStartRecord(
  payload: Uint8Array<ArrayBufferLike>
): TerminalMultiplexSnapshotStartRecord | null {
  const wire = TERMINAL_MULTIPLEX_SNAPSHOT_RECORD_WIRE.start
  if (payload.byteLength !== wire.bytes) {
    return null
  }
  const view = payloadView(payload)
  const reasonWire = view.getUint8(wire.reasonOffset)
  const reason = SNAPSHOT_REASON_BY_WIRE[reasonWire]
  const source = view.getUint8(wire.sourceOffset)
  const activeBuffer = view.getUint8(wire.activeBufferOffset)
  const flags = view.getUint8(wire.flagsOffset)
  const coverageEndSeq = view.getBigUint64(wire.coverageEndSeqOffset, true)
  const pendingDeliveryStartSeq = view.getBigUint64(wire.pendingDeliveryStartSeqOffset, true)
  if (
    reason === undefined ||
    reasonWire > wire.reasonMax ||
    source > wire.sourceMax ||
    activeBuffer > wire.activeBufferMax ||
    (flags & ~wire.flagsMask) !== 0 ||
    view.getUint16(wire.colsOffset, true) < wire.colsMin ||
    view.getUint16(wire.colsOffset, true) > wire.colsMax ||
    view.getUint16(wire.rowsOffset, true) < wire.rowsMin ||
    view.getUint16(wire.rowsOffset, true) > wire.rowsMax ||
    view.getBigUint64(wire.reserved64Offset, true) !== 0n ||
    pendingDeliveryStartSeq > coverageEndSeq ||
    view.getUint32(wire.reserved32Offset, true) !== 0
  ) {
    return null
  }
  return {
    snapshotId: view.getUint32(wire.snapshotIdOffset, true),
    reason,
    source: source as TerminalMultiplexSnapshotStartRecord['source'],
    activeBuffer: activeBuffer as TerminalMultiplexSnapshotStartRecord['activeBuffer'],
    truncated: (flags & wire.truncatedFlag) !== 0,
    byteBudget: (flags & wire.byteBudgetFlag) !== 0,
    coldRestore: (flags & wire.coldRestoreFlag) !== 0,
    cols: view.getUint16(wire.colsOffset, true),
    rows: view.getUint16(wire.rowsOffset, true),
    retainedScrollbackRows: view.getUint32(wire.retainedScrollbackRowsOffset, true),
    coverageEndSeq,
    pendingDeliveryStartSeq,
    sectionBytes: [
      view.getUint32(wire.sectionBytesOffset, true),
      view.getUint32(wire.sectionBytesOffset + wire.sectionStrideBytes, true),
      view.getUint32(wire.sectionBytesOffset + wire.sectionStrideBytes * 2, true),
      view.getUint32(wire.sectionBytesOffset + wire.sectionStrideBytes * 3, true),
      view.getUint32(wire.sectionBytesOffset + wire.sectionStrideBytes * 4, true)
    ]
  }
}

export function encodeTerminalMultiplexSnapshotReason(
  reason: TerminalMultiplexSnapshotReason
): number {
  return SNAPSHOT_REASON_BY_WIRE.indexOf(reason)
}

export function encodeTerminalMultiplexSnapshotChunkRecord(
  record: TerminalMultiplexSnapshotChunkRecord
): Uint8Array<ArrayBuffer> {
  const wire = TERMINAL_MULTIPLEX_SNAPSHOT_RECORD_WIRE.chunk
  const out = new Uint8Array(wire.headerBytes + record.data.byteLength)
  const view = new DataView(out.buffer)
  view.setUint32(wire.snapshotIdOffset, record.snapshotId, true)
  view.setUint8(wire.sectionOffset, record.section)
  view.setUint8(wire.reserved8Offset, 0)
  view.setUint16(wire.reserved16Offset, 0, true)
  view.setUint32(wire.dataOffsetOffset, record.sectionOffset, true)
  view.setUint32(wire.dataBytesOffset, record.data.byteLength, true)
  out.set(record.data, wire.headerBytes)
  return out
}

export function decodeTerminalMultiplexSnapshotChunkRecord(
  payload: Uint8Array<ArrayBufferLike>
): TerminalMultiplexSnapshotChunkRecord | null {
  const wire = TERMINAL_MULTIPLEX_SNAPSHOT_RECORD_WIRE.chunk
  if (payload.byteLength < wire.headerBytes) {
    return null
  }
  const view = payloadView(payload)
  const section = view.getUint8(wire.sectionOffset)
  if (
    section > wire.sectionMax ||
    view.getUint8(wire.reserved8Offset) !== 0 ||
    view.getUint16(wire.reserved16Offset, true) !== 0 ||
    view.getUint32(wire.dataBytesOffset, true) !== payload.byteLength - wire.headerBytes ||
    payload.byteLength - wire.headerBytes > wire.maxDataBytes
  ) {
    return null
  }
  return {
    snapshotId: view.getUint32(wire.snapshotIdOffset, true),
    section: section as TerminalMultiplexSnapshotChunkRecord['section'],
    sectionOffset: view.getUint32(wire.dataOffsetOffset, true),
    data: payload.slice(wire.headerBytes)
  }
}

export function encodeTerminalMultiplexSnapshotEndRecord(
  record: TerminalMultiplexSnapshotEndRecord
): Uint8Array<ArrayBuffer> {
  const wire = TERMINAL_MULTIPLEX_SNAPSHOT_RECORD_WIRE.end
  const out = new Uint8Array(wire.bytes)
  const view = new DataView(out.buffer)
  view.setUint32(wire.snapshotIdOffset, record.snapshotId, true)
  view.setUint8(wire.statusOffset, record.status)
  view.setUint8(wire.reserved8Offset, 0)
  view.setUint16(wire.reserved16Offset, 0, true)
  view.setBigUint64(wire.coverageEndSeqOffset, record.coverageEndSeq, true)
  view.setUint32(wire.assembledBytesOffset, record.assembledBytes, true)
  view.setUint32(wire.crc32cOffset, record.crc32c, true)
  return out
}

export function decodeTerminalMultiplexSnapshotEndRecord(
  payload: Uint8Array<ArrayBufferLike>
): TerminalMultiplexSnapshotEndRecord | null {
  const wire = TERMINAL_MULTIPLEX_SNAPSHOT_RECORD_WIRE.end
  if (payload.byteLength !== wire.bytes) {
    return null
  }
  const view = payloadView(payload)
  const status = view.getUint8(wire.statusOffset)
  if (
    status > wire.statusMax ||
    view.getUint8(wire.reserved8Offset) !== 0 ||
    view.getUint16(wire.reserved16Offset, true) !== 0
  ) {
    return null
  }
  return {
    snapshotId: view.getUint32(wire.snapshotIdOffset, true),
    status: status as TerminalMultiplexSnapshotEndRecord['status'],
    coverageEndSeq: view.getBigUint64(wire.coverageEndSeqOffset, true),
    assembledBytes: view.getUint32(wire.assembledBytesOffset, true),
    crc32c: view.getUint32(wire.crc32cOffset, true)
  }
}

function payloadView(payload: Uint8Array<ArrayBufferLike>): DataView {
  return new DataView(payload.buffer, payload.byteOffset, payload.byteLength)
}

File: apps/mobile-ios/YiruMobile/Platform/Terminal/Multiplex/SnapshotRecords.swift
import Foundation

nonisolated struct TerminalMultiplexSnapshotStartRecord: Equatable, Sendable {
    let snapshotID: UInt32
    let reason: UInt8
    let source: UInt8
    let activeBuffer: UInt8
    let isTruncated: Bool
    let isByteBudgetLimited: Bool
    let isColdRestore: Bool
    let columns: UInt16
    let rows: UInt16
    let retainedScrollbackRows: UInt32
    let coverageEndSequence: UInt64
    let pendingDeliveryStartSequence: UInt64
    let sectionByteCounts: [UInt32]
}

nonisolated struct TerminalMultiplexSnapshotChunkRecord: Equatable, Sendable {
    let snapshotID: UInt32
    let section: UInt8
    let sectionOffset: UInt32
    let data: Data
}

nonisolated struct TerminalMultiplexSnapshotEndRecord: Equatable, Sendable {
    let snapshotID: UInt32
    let status: UInt8
    let coverageEndSequence: UInt64
    let assembledBytes: UInt32
    let crc32c: UInt32
}

nonisolated enum TerminalMultiplexSnapshotRecordError: Error, Equatable, Sendable {
    case invalidStart
    case invalidChunk
    case invalidEnd
}

nonisolated enum TerminalMultiplexSnapshotRecordCodec {
    static let chunkDataBytes = TerminalMultiplexSnapshotChunkRecordWire.maxDataBytes

    static func decodeStart(_ data: Data) throws -> TerminalMultiplexSnapshotStartRecord {
        let wire = TerminalMultiplexSnapshotStartRecordWire.self
        guard data.count == wire.bytes else {
            throw TerminalMultiplexSnapshotRecordError.invalidStart
        }
        let reason = TerminalWireBytes.byte(in: data, at: wire.reasonOffset)
        let source = TerminalWireBytes.byte(in: data, at: wire.sourceOffset)
        let activeBuffer = TerminalWireBytes.byte(in: data, at: wire.activeBufferOffset)
        let flags = TerminalWireBytes.byte(in: data, at: wire.flagsOffset)
        let columns = TerminalWireBytes.uint16(in: data, at: wire.colsOffset)
        let rows = TerminalWireBytes.uint16(in: data, at: wire.rowsOffset)
        let coverage = TerminalWireBytes.uint64(in: data, at: wire.coverageEndSeqOffset)
        let pendingDelivery = TerminalWireBytes.uint64(
            in: data,
            at: wire.pendingDeliveryStartSeqOffset
        )
        guard Int(reason) <= wire.reasonMax, Int(source) <= wire.sourceMax,
            Int(activeBuffer) <= wire.activeBufferMax, Int(flags) & ~wire.flagsMask == 0,
            (wire.colsMin...wire.colsMax).contains(Int(columns)),
            (wire.rowsMin...wire.rowsMax).contains(Int(rows)),
            TerminalWireBytes.uint64(in: data, at: wire.reserved64Offset) == 0,
            pendingDelivery <= coverage,
            TerminalWireBytes.uint32(in: data, at: wire.reserved32Offset) == 0
        else {
            throw TerminalMultiplexSnapshotRecordError.invalidStart
        }
        return TerminalMultiplexSnapshotStartRecord(
            snapshotID: TerminalWireBytes.uint32(in: data, at: wire.snapshotIdOffset),
            reason: reason,
            source: source,
            activeBuffer: activeBuffer,
            isTruncated: Int(flags) & wire.truncatedFlag != 0,
            isByteBudgetLimited: Int(flags) & wire.byteBudgetFlag != 0,
            isColdRestore: Int(flags) & wire.coldRestoreFlag != 0,
            columns: columns,
            rows: rows,
            retainedScrollbackRows: TerminalWireBytes.uint32(
                in: data,
                at: wire.retainedScrollbackRowsOffset
            ),
            coverageEndSequence: coverage,
            pendingDeliveryStartSequence: pendingDelivery,
            sectionByteCounts: (0..<wire.sectionCount).map {
                TerminalWireBytes.uint32(
                    in: data,
                    at: wire.sectionBytesOffset + $0 * wire.sectionStrideBytes
                )
            }
        )
    }

    static func decodeChunk(_ data: Data) throws -> TerminalMultiplexSnapshotChunkRecord {
        let wire = TerminalMultiplexSnapshotChunkRecordWire.self
        guard data.count >= wire.headerBytes else {
            throw TerminalMultiplexSnapshotRecordError.invalidChunk
        }
        let section = TerminalWireBytes.byte(in: data, at: wire.sectionOffset)
        let byteCount = Int(TerminalWireBytes.uint32(in: data, at: wire.dataBytesOffset))
        guard Int(section) <= wire.sectionMax,
            TerminalWireBytes.byte(in: data, at: wire.reserved8Offset) == 0,
            TerminalWireBytes.uint16(in: data, at: wire.reserved16Offset) == 0,
            byteCount == data.count - wire.headerBytes,
            byteCount <= chunkDataBytes
        else {
            throw TerminalMultiplexSnapshotRecordError.invalidChunk
        }
        return TerminalMultiplexSnapshotChunkRecord(
            snapshotID: TerminalWireBytes.uint32(in: data, at: wire.snapshotIdOffset),
            section: section,
            sectionOffset: TerminalWireBytes.uint32(in: data, at: wire.dataOffsetOffset),
            data: data.subdata(in: wire.headerBytes..<data.count)
        )
    }

    static func decodeEnd(_ data: Data) throws -> TerminalMultiplexSnapshotEndRecord {
        let wire = TerminalMultiplexSnapshotEndRecordWire.self
        guard data.count == wire.bytes,
            Int(TerminalWireBytes.byte(in: data, at: wire.statusOffset)) <= wire.statusMax,
            TerminalWireBytes.byte(in: data, at: wire.reserved8Offset) == 0,
            TerminalWireBytes.uint16(in: data, at: wire.reserved16Offset) == 0
        else {
            throw TerminalMultiplexSnapshotRecordError.invalidEnd
        }
        return TerminalMultiplexSnapshotEndRecord(
            snapshotID: TerminalWireBytes.uint32(in: data, at: wire.snapshotIdOffset),
            status: TerminalWireBytes.byte(in: data, at: wire.statusOffset),
            coverageEndSequence: TerminalWireBytes.uint64(
                in: data,
                at: wire.coverageEndSeqOffset
            ),
            assembledBytes: TerminalWireBytes.uint32(
                in: data,
                at: wire.assembledBytesOffset
            ),
            crc32c: TerminalWireBytes.uint32(in: data, at: wire.crc32cOffset)
        )
    }
}
